File: CSES/1. Introductory Problems/CSES-1624_Chessboard and Queens.py
""" CSES 1624 - Chessboard and Queens
    Backtracking
"""
n = 8
mp = [input() for _ in range(n)]
path = [0] * n #  row-i 的 Q 在 col-path[i]

ans = 0
def check(x, y): 
    for i in range(x): # 檢查前 x row
        j = path[i]
        if (x+y) == (i+j) or (x-y) == (i-j): # 位在同一條斜線上 (斜率相同)
            return False
    return True

def dfs(r, avail):
    global ans
    if r == n:
        ans += 1
        return
    for c in avail:
        if mp[r][c] == "." and check(r, c):
            path[r] = c
            dfs(r+1, avail-{c})
dfs(0, set(range(n)))
print(ans)

File: CSES/1. Introductory Problems/CSES-1624_Chessboard and Queens.cpp
/* CSES 1624: Chessboard and Queens
    Backtracking + Bitmask
*/
#include <bits/stdc++.h>
using namespace std;
const int N = 8;
#define endl '\n'

vector<string> mp(N);
vector<int> path(N); // row-i 的 Q 在 col-path[i]
int ans = 0;

bool check(int x, int y) {
    for (int i = 0; i < x; i++) {
        int j = path[i];
        if ((x+y) == (i+j) || (x-y) == (i-j)) return false; // 位在同一條斜線上 (斜率相同)
    }
    return true;
}

void dfs(int r, int avail) {
    if (r == N) {
        ans++;
        return;
    }
    for (int c = 0; c < N; c++) {
        if (avail & (1 << c) && mp[r][c] == '.' && check(r, c)) {
            path[r] = c;
            dfs(r+1, avail & ~(1 << c));
        }
    }
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    for (int i = 0; i < N; i++) cin >> mp[i];
    dfs(0, (1 << N) - 1);
    cout << ans << endl;
    return 0;
}